ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,270 @@
+    from graphql.language.ast import Document
+else:
+    from graphql.language.ast import DocumentNode as Document
+
+
+config._add(
+    "graphql",
+    dict(
+        _default_service="graphql",
+        resolvers_enabled=asbool(os.getenv("DD_TRACE_GRAPHQL_RESOLVERS_ENABLED", default=False)),
+    ),
+)
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False):
+        return
+    setattr(graphql, "_datadog_patch", True)
+    Pin().onto(graphql)
+
+    for module_str, func_name, wrapper in _get_patching_candidates():
+        _update_patching(wrap, module_str, func_name, wrapper)
+
+
+def unpatch():
+    if not getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+
+    for module_str, func_name, wrapper in _get_patching_candidates():
+        _update_patching(unwrap, module_str, func_name, wrapper)
+
+    setattr(graphql, "_datadog_patch", False)
+
+
+def _get_patching_candidates():
+    if graphql_version < (3, 0):
+        return [
+            ("graphql.graphql", "execute_graphql", _traced_query),
+            ("graphql.language.parser", "parse", _traced_parse),
+            ("graphql.validation.validation", "validate", _traced_validate),
+            ("graphql.execution.executor", "execute", _traced_execute),
+        ]
+    return [
+        ("graphql.graphql", "graphql_impl", _traced_query),
+        ("graphql.language.parser", "parse", _traced_parse),
+        ("graphql.validation.validate", "validate", _traced_validate),
+        ("graphql.execution.execute", "execute", _traced_execute),
+    ]
+
+
+def _update_patching(operation, module_str, func_name, wrapper):
+    module = sys.modules[module_str]
+    func = getattr(module, func_name)
+    operation(func, wrapper)
+
+
+def _traced_parse(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    with pin.tracer.trace(
+        name="graphql.parse",
+        service=trace_utils.int_service(pin, config.graphql),
+        span_type=SpanTypes.GRAPHQL,
+    ) as span:
+        _init_span(span)
+        return func(*args, **kwargs)
+
+
+def _traced_validate(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    with pin.tracer.trace(
+        name="graphql.validate",
+        service=trace_utils.int_service(pin, config.graphql),
+        span_type=SpanTypes.GRAPHQL,
+    ) as span:
+        _init_span(span)
+        errors = func(*args, **kwargs)
+        _set_span_errors(errors, span)
+        return errors
+
+
+def _traced_execute(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    if config.graphql.resolvers_enabled:
+        # patch resolvers
+        args, kwargs = _inject_trace_middleware_to_args(_resolver_middleware, args, kwargs)
+
+    # set resource name
+    if graphql_version < (3, 0):
```

Comment:
Probably not a biggie, but I think it would be grand if there was a way to avoid branching. For example, here we could precompute the kwarg name after we set the global `graphql_version` and use that instead. We would still be doing a global lookup, but save the branching. Something similar could be down a few lines down as well. WDYT?

Reply:
There's a few areas where we branch to handle different graphql versions. Creating global variables for all might come with some performance gains but it would make the integration harder to maintain